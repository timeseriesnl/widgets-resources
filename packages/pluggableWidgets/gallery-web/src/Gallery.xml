<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.gallery.Gallery" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Gallery</name>
    <description/>
    <icon/>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="datasource" type="datasource" isList="true">
                    <caption>Data source</caption>
                    <description/>
                </property>
                <property key="content" type="widgets" dataSource="datasource">
                    <caption>Content</caption>
                    <description/>
                </property>
                <property key="desktopItems" type="integer" defaultValue="4">
                    <caption>Desktop items</caption>
                    <description />
                </property>
                <property key="tabletItems" type="integer" defaultValue="3">
                    <caption>Tablet items</caption>
                    <description />
                </property>
                <property key="phoneItems" type="integer" defaultValue="2">
                    <caption>Phone items</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Items">
                <property key="pageSize" type="integer" defaultValue="20">
                    <caption>Page size</caption>
                    <description />
                </property>
                <property key="pagination" type="enumeration" defaultValue="buttons">
                    <caption>Pagination</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="buttons">Paging buttons</enumerationValue>
                        <enumerationValue key="virtualScrolling">Virtual scrolling</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="pagingPosition" type="enumeration" defaultValue="below">
                    <caption>Position of paging buttons</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="below">Below grid</enumerationValue>
                        <enumerationValue key="above">Above grid</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="showEmptyPlaceholder" type="enumeration" defaultValue="none">
                    <caption>Empty message</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="emptyPlaceholder" type="widgets" required="false">
                    <caption>Empty placeholder</caption>
                    <description/>
                </property>
                <property key="itemClass" type="expression" required="false" dataSource="datasource">
                    <caption>Dynamic item class</caption>
                    <description />
                    <returnType type="String"/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onClick" type="action" required="false" dataSource="datasource">
                    <caption>On click action</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
