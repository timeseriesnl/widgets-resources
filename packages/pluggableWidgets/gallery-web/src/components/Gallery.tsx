import { createElement, ReactElement, ReactNode } from "react";
import classNames from "classnames";

export interface GalleryProps<T> {
    className?: string;
    desktopItems: number;
    items: T[];
    itemRenderer: (
        renderWrapper: (children: ReactNode, className?: string, onClick?: () => void) => ReactElement,
        item: T
    ) => ReactNode;
    phoneItems: number;
    tabletItems: number;
    tabIndex?: number;
}

export function Gallery<T>(props: GalleryProps<T>): ReactElement {
    return (
        <div
            className={classNames(
                "widget-gallery",
                props.className,
                `widget-gallery-lg-${props.desktopItems}`,
                `widget-gallery-md-${props.tabletItems}`,
                `widget-gallery-sm-${props.phoneItems}`
            )}
            data-focusindex={props.tabIndex || 0}
        >
            {props.items.map(item =>
                props.itemRenderer((children, className, onClick) => {
                    return (
                        <div
                            key={`item_${item}`}
                            className={classNames("widget-gallery-item", className, {
                                "widget-gallery-clickable": !!onClick
                            })}
                            onClick={onClick}
                            onKeyDown={
                                onClick
                                    ? e => {
                                          if (e.key === "Enter" || e.key === " ") {
                                              e.preventDefault();
                                              onClick();
                                          }
                                      }
                                    : undefined
                            }
                            role={onClick ? "button" : undefined}
                            tabIndex={onClick ? 0 : undefined}
                        >
                            {children}
                        </div>
                    );
                }, item)
            )}
        </div>
    );
}
