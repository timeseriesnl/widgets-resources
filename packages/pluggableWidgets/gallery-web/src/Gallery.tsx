import { createElement, ReactElement, useCallback } from "react";
import { GalleryContainerProps } from "../typings/GalleryProps";
import { Gallery as GalleryComponent } from "./components/Gallery";
import "./ui/gallery-main.scss";

export function Gallery(props: GalleryContainerProps): ReactElement {
    return (
        <GalleryComponent
            className={props.class}
            desktopItems={props.desktopItems}
            items={props.datasource.items ?? []}
            itemRenderer={useCallback(
                (renderWrapper, item) =>
                    renderWrapper(
                        props.content.get(item),
                        props.itemClass?.get(item)?.value,
                        props.onClick ? props.onClick?.get(item).execute : undefined
                    ),
                [props.content, props.itemClass, props.onClick]
            )}
            phoneItems={props.phoneItems}
            tabletItems={props.tabletItems}
            tabIndex={props.tabIndex}
        />
    );
}
