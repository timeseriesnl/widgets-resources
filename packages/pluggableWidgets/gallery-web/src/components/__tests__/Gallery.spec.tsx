import { createElement } from "react";
import { mount, shallow } from "enzyme";
import { Gallery, GalleryProps } from "../Gallery";

const itemWrapperFunction = ({
    onClick,
    customClass
}: {
    onClick?: () => void;
    customClass?: string;
}): GalleryProps<string>["itemRenderer"] => (wrapper, item) => wrapper(item, customClass, onClick);

const defaultProps: GalleryProps<string> = {
    phoneItems: 2,
    tabletItems: 3,
    desktopItems: 4,
    className: "",
    items: ["11", "22", "33"],
    itemRenderer: itemWrapperFunction({})
};

describe("Gallery", () => {
    describe("DOM Structure", () => {
        it("renders correctly", () => {
            const gallery = shallow(<Gallery {...defaultProps} />);

            expect(gallery).toMatchSnapshot();
        });

        it("renders correctly with onclick event", () => {
            const gallery = shallow(
                <Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick: jest.fn() })} />
            );

            expect(gallery).toMatchSnapshot();
        });
    });

    describe("with events", () => {
        it("triggers correct events on click", () => {
            const onClick = jest.fn();
            const gallery = mount(<Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick })} />);
            const galleryFirstItem = gallery.find(".widget-gallery-clickable").at(0);

            expect(galleryFirstItem).toBeDefined();

            galleryFirstItem.simulate("click");

            expect(onClick).toBeCalled();
        });

        it("triggers correct events on Enter key down", () => {
            const onClick = jest.fn();
            const gallery = mount(<Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick })} />);
            const galleryFirstItem = gallery.find(".widget-gallery-clickable").at(0);

            expect(galleryFirstItem).toBeDefined();

            galleryFirstItem.simulate("keydown", { key: "Enter" });

            expect(onClick).toBeCalled();
        });

        it("triggers correct events on Space key down", () => {
            const onClick = jest.fn();
            const gallery = mount(<Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ onClick })} />);
            const galleryFirstItem = gallery.find(".widget-gallery-clickable").at(0);

            expect(galleryFirstItem).toBeDefined();

            galleryFirstItem.simulate("keydown", { key: " " });

            expect(onClick).toBeCalled();
        });
    });

    describe("with different configurations per platform", () => {
        it("contains correct classes for desktop", () => {
            const gallery = shallow(<Gallery {...defaultProps} desktopItems={12} />);

            expect(gallery.hasClass("widget-gallery-lg-12")).toBeTruthy();
        });

        it("contains correct classes for tablet", () => {
            const gallery = shallow(<Gallery {...defaultProps} tabletItems={6} />);

            expect(gallery.hasClass("widget-gallery-md-6")).toBeTruthy();
        });

        it("contains correct classes for phone", () => {
            const gallery = shallow(<Gallery {...defaultProps} phoneItems={3} />);

            expect(gallery.hasClass("widget-gallery-sm-3")).toBeTruthy();
        });
    });

    describe("with custom classes", () => {
        it("contains correct classes in the wrapper", () => {
            const gallery = shallow(<Gallery {...defaultProps} className="custom-class" />);

            expect(gallery.hasClass("custom-class")).toBeTruthy();
        });

        it("contains correct classes in the items", () => {
            const gallery = shallow(
                <Gallery {...defaultProps} itemRenderer={itemWrapperFunction({ customClass: "custom-class" })} />
            );
            const galleryFirstItem = gallery.find(".widget-gallery-item").at(0);

            expect(galleryFirstItem.hasClass("custom-class")).toBeTruthy();
        });
    });
});
