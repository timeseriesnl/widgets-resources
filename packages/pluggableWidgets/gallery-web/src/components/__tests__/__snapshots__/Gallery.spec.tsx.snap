// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery DOM Structure renders correctly 1`] = `
<div
  className="widget-gallery widget-gallery-lg-4 widget-gallery-md-3 widget-gallery-sm-2"
  data-focusindex={0}
>
  <div
    className="widget-gallery-item"
    key="item_11"
  >
    11
  </div>
  <div
    className="widget-gallery-item"
    key="item_22"
  >
    22
  </div>
  <div
    className="widget-gallery-item"
    key="item_33"
  >
    33
  </div>
</div>
`;

exports[`Gallery DOM Structure renders correctly with onclick event 1`] = `
<div
  className="widget-gallery widget-gallery-lg-4 widget-gallery-md-3 widget-gallery-sm-2"
  data-focusindex={0}
>
  <div
    className="widget-gallery-item widget-gallery-clickable"
    key="item_11"
    onClick={[MockFunction]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    11
  </div>
  <div
    className="widget-gallery-item widget-gallery-clickable"
    key="item_22"
    onClick={[MockFunction]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    22
  </div>
  <div
    className="widget-gallery-item widget-gallery-clickable"
    key="item_33"
    onClick={[MockFunction]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    33
  </div>
</div>
`;
